### PR TITLE
Add guess to dataclass transformer

### PR DESCRIPTION
**Pull request by pingsutw**
Signed-off-by: Kevin Su <pingsutw@apache.org>

# TL;DR
Convert dataclass JSON schema to a python class

## Type
 - [x] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [ ] Code completed
 - [ ] Smoke tested
 - [x] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
https://flyte-org.slack.com/archives/CREL4QVAQ/p1632423587185900

## Follow-up issue
_NA_

### CODE REVIEW

**Comment by pingsutw** [comment]
cc @kumare3 @eapolinario

File: setup.py [hunk 1]
```diff
@@ -94,6 +94,7 @@
         "singledispatchmethod; python_version < '3.8.0'",
         "docstring-parser>=0.9.0",
         "diskcache>=5.2.1",
+        "warlock>=1.3.3",
```
Comment (kumare3) [inline setup.py:97]: Do we want to add one more dependency?

Comment (kumare3) [reply setup.py:97]: What I mean - https://pypi.org/project/warlock/ - no python 3.8?

Comment (pingsutw) [reply setup.py:97]: Maybe I can try to implement it by myself.
https://github.com/bcwaldon/warlock/blob/6f41194ab675efa451470f8abfebfa6bc72403fe/warlock/core.py#L23

File: flytekit/core/type_engine.py [hunk 2]
```diff
@@ -731,6 +734,30 @@ def to_python_value(self, ctx: FlyteContext, lv: Literal, expected_python_type:
...
+    """
+    schema = copy.deepcopy(schema)
+
+    class Model(dict):
+        def __init__(self, *args, **kwargs):
+            self.__dict__["schema"] = schema
+            d = dict(*args, **kwargs)
+            dict.__init__(self, d)
+
+        def __setitem__(self, key, value):
```
Comment (kumare3) [inline flytekit/core/type_engine.py:749]: why should we support setitem?

File: flytekit/core/type_engine.py [hunk 3]
```diff
@@ -731,6 +734,30 @@ def to_python_value(self, ctx: FlyteContext, lv: Literal, expected_python_type:
...
+            dict.__setitem__(self, key, value)
+
+        def __getattr__(self, key):
+            try:
+                return self.__getitem__(key)
```
Comment (kumare3) [inline flytekit/core/type_engine.py:754]: what if a nested item is a dataclass / schema too?

File: tests/flytekit/unit/core/test_type_hints.py [hunk 4]
```diff
@@ -1342,6 +1342,30 @@ def t2() -> dict:
     assert output_lm.literals["o0"].scalar.generic == expected_struct
 
 
+def test_guess_dict4():
```
Comment (kumare3) [inline tests/flytekit/unit/core/test_type_hints.py:1345]: lets have an example of nested dataclass

```
@dataclass
class Nested():
   ...

@dataclass
class Outer():
    n: Nested
```

Comment (pingsutw) [reply tests/flytekit/unit/core/test_type_hints.py:1345]: sure

File: flytekit/core/type_engine.py (hunk 2 above)
Comment (pingsutw) [reply flytekit/core/type_engine.py:749]: yeah, we don't use it now. I will remove it, and we can add it in the future if we need it.

File: flytekit/core/type_engine.py (hunk 3 above)
Comment (pingsutw) [reply flytekit/core/type_engine.py:751]: It also works, check the new test below.

**Comment by kumare3** [comment]
This is cool awesome. I thought about it some more, let's add set item- use case - retrieve results, modify and resubmit a new run in flytekit remote

**Review by kumare3 — APPROVED** [review]

**Pull request merged**